### PR TITLE
diagram builder: introduce AddNamedSystem sugar

### LINKED ISSUE #5895: Should Modify DiagramBuilder::AddSystem() to take the name of the system as a parameter.
This is API sugar to avoid the common idiom of creating a system and then immediately naming it. For example, instead of:

```
auto pursuit = builder_->template AddSystem<PurePursuitController<T>>(*road_);
pursuit->set_name("foo");
```
One would type:

```
auto pursuit = builder_->template AddSystem<PurePursuitController<T>>("foo", *road_);
```

Alternatively:

```
auto pursuit = builder_->template AddNamedSystem<PurePursuitController<T>>("foo", *road_);
```

Maybe `AddSystem()` can fail-fast if the provided name is not unique within the to-be-created `Diagram`.

### PR DESCRIPTION
**Pull request by rpoyner-tri**
Closes #5895.

This patch provides an potentially more fluent alternative to the common
AddSystem/set_name idiom.

---
This change is [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/14404)

### CODE REVIEW

**Review by rpoyner-tri — COMMENTED** [review]
+@ericcousineau-tri for feature, since this touches bindings, and the open memory leak issue in particular.

> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/14404#-:-MNZlL0e2oKYwhARoLBz:b-b0ozdi)* status: LGTM missing from assignee EricCousineau-TRI(platform), needs at least two assigned reviewers (waiting on @EricCousineau-TRI)

**Review by EricCousineau-TRI — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/14404)* status: 1 unresolved discussion, LGTM missing from assignee EricCousineau-TRI(platform), needs at least two assigned reviewers (waiting on @EricCousineau-TRI and @rpoyner-tri)

*[a discussion](https://reviewable.io/reviews/robotlocomotion/drake/14404#-MN_IYeCDR0J27_9AM32:-MN_IYeED_yK6qiR7kaD:b9gcw70) (no related file):*
Sorry, I may not be the best feature reviewer. I'm not a stakeholder in this sugar, and I dislike overly coupled interfaces for the sake of simple sugar.
(I prefer the `AddSystem/set_name` idiom because it's more clearly separated, and you don't need to repeat weird binding patterns.)

If you'd like, I can accept your motivation blindly and still allow you to move forward.

Is that what you would like?

---

**Review by rpoyner-tri — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/14404)* status: 1 unresolved discussion, LGTM missing from assignee EricCousineau-TRI(platform), needs at least two assigned reviewers (waiting on @EricCousineau-TRI)

*[a discussion](https://reviewable.io/reviews/robotlocomotion/drake/14404#-MN_IYeCDR0J27_9AM32:-MN_KrAe94bgpu5yux90:b-6fjk01) (no related file):*
<details><summary><i>Previously, EricCousineau-TRI (Eric Cousineau) wrote…</i></summary><blockquote>

Sorry, I may not be the best feature reviewer. I'm not a stakeholder in this sugar, and I dislike overly coupled interfaces for the sake of simple sugar.
(I prefer the `AddSystem/set_name` idiom because it's more clearly separated, and you don't need to repeat weird binding patterns.)

If you'd like, I can accept your motivation blindly and still allow you to move forward.

Is that what you would like?
</blockquote></details>

I'd say there are two options, since I am blindly accepting the motivation of #5895:

1. join me in drinking the kool-aid.
2. Convince me that I should just close the issue wontfix and get on with my life.

Either outcome is acceptable to me, but i would like to have good arguments for them.

META: Drake's issue base is littered with issues that have been argued about but never resolved; my method is to select one or more theories and try them out in hopes of forcing resolution.

---

**Comment by sherm1** [comment]
FWIW I'd be happy to Wontfix this one and move on to #7128 so that the default names are reasonable.

**Review by EricCousineau-TRI — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/14404)* status: 1 unresolved discussion, LGTM missing from assignee EricCousineau-TRI(platform), needs at least two assigned reviewers (waiting on @EricCousineau-TRI and @sherm1)

*[a discussion](https://reviewable.io/reviews/robotlocomotion/drake/14404#-MN_IYeCDR0J27_9AM32:-MN_d6Mu47LK6mTxq3_T:b-oea10s) (no related file):*
> Convince me that I should just close the issue wontfix and get on with my life.

Sherm's allusion to working on #7128 may be the best proposition. I'm not entirely sure where that stands in terms of the current codebase / stakeholders, tho.

@sherm1 Can you expand on the status of that issue w.r.t. the current code base?
(like, we still have `0xbadf00d` like numbers in our system names, which is hella non-interperetable?)

> META: Drake's issue base is littered with issues that have been argued about but never resolved; my method is to select one or more theories and try them out in hopes of forcing resolution.

Makes sense!
My guesstimate is that Jeremy may still provide this PR as it's rendered, but I can't say.
Would you be OK with waiting until you get his input?

I'd prefer someone have a slightly-more-than-neutral-positive opinion on this PR.
As it stands, none of us three fit the bill? (and it's public API that has had some recent weird churn for Python land :( )

---

On a separate thread, I *do* have something that I'd love to see more:
Fewer `DeclareMyEventTypeShindig(PublishOrWhateverEventDataThatMayBeOpaquelyRedundant)` things!!!
https://github.com/RobotLocomotion/drake/issues/13165

---

TL;DR I'm still happy to review it if the response is "too much meta convo, let's move forward plz".

---

**Review by sherm1 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/14404)* status: 1 unresolved discussion, LGTM missing from assignee EricCousineau-TRI(platform), needs at least two assigned reviewers (waiting on @EricCousineau-TRI)

*[a discussion](https://reviewable.io/reviews/robotlocomotion/drake/14404#-MN_IYeCDR0J27_9AM32:-MN_vEzl-m1nHKltgGjo:b-41drp7) (no related file):*
<details><summary><i>Previously, EricCousineau-TRI (Eric Cousineau) wrote…</i></summary><blockquote>

> Convince me that I should just close the issue wontfix and get on with my life.

Sherm's allusion to working on #7128 may be the best proposition. I'm not entirely sure where that stands in terms of the current codebase / stakeholders, tho.

@sherm1 Can you expand on the status of that issue w.r.t. the current code base?
(like, we still have `0xbadf00d` like numbers in our system names, which is hella non-interperetable?)

> META: Drake's issue base is littered with issues that have been argued about but never resolved; my method is to select one or more theories and try them out in hopes of forcing resolution.

Makes sense!
My guesstimate is that Jeremy may still provide this PR as it's rendered, but I can't say.
Would you be OK with waiting until you get his input?

I'd prefer someone have a slightly-more-than-neutral-positive opinion on this PR.
As it stands, none of us three fit the bill? (and it's public API that has had some recent weird churn for Python land :( )

---

On a separate thread, I *do* have something that I'd love to see more:
Fewer `DeclareMyEventTypeShindig(PublishOrWhateverEventDataThatMayBeOpaquelyRedundant)` things!!!
https://github.com/RobotLocomotion/drake/issues/13165

---

TL;DR I'm still happy to review it if the response is "too much meta convo, let's move forward plz".
</blockquote></details>

@ericcousineau-tri I think #7128 is still needed and would get rid of that awful gobbledygook in the generated names we have. I doubt the current scheme has any enthusiastic supporters! I also think the change to clean names would be painless since clearly no one is depending on the specifics of the hex-filled unpredictable generated names now.

---

**Review by EricCousineau-TRI — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/14404)* status: 1 unresolved discussion, LGTM missing from assignee EricCousineau-TRI(platform), needs at least two assigned reviewers (waiting on @EricCousineau-TRI, @rpoyner-tri, and @sherm1)

*[a discussion](https://reviewable.io/reviews/robotlocomotion/drake/14404#-MN_IYeCDR0J27_9AM32:-MNa0W2lENET5Xm6eTnm:b-cyxwtp) (no related file):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

@ericcousineau-tri I think #7128 is still needed and would get rid of that awful gobbledygook in the generated names we have. I doubt the current scheme has any enthusiastic supporters! I also think the change to clean names would be painless since clearly no one is depending on the specifics of the hex-filled unpredictable generated names now.
</blockquote></details>

Sweet, then no argument here!

So my takeaway is:

- #7128 has buyers present, and seems heckin' useful.
- This PR (for #5895) does not (yet); may still be useful tho, so worth not closing?

@rpoyner-tri You agree with this sentiment?

---

**Review by avalenzu — COMMENTED** [review]
-@ericcousineau-tri +@avalenzu for feature review. I would definitely like to have this.

> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/14404#-:-MNa3Irr8PjRpjVhIFuC:b689xyy)* status: 1 unresolved discussion, LGTM missing from assignee avalenzu, needs platform reviewer assigned, needs at least two assigned reviewers (waiting on @avalenzu, @rpoyner-tri, and @sherm1)

**Review by avalenzu — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/14404)* status: 1 unresolved discussion, LGTM missing from assignee avalenzu, needs platform reviewer assigned, needs at least two assigned reviewers (waiting on @avalenzu and @EricCousineau-TRI)

*[a discussion](https://reviewable.io/reviews/robotlocomotion/drake/14404#-MN_IYeCDR0J27_9AM32:-MNa3T8l60wDlOR3XvWJ:b-nmz3dt) (no related file):*
<details><summary><i>Previously, EricCousineau-TRI (Eric Cousineau) wrote…</i></summary><blockquote>

Sweet, then no argument here!

So my takeaway is:

- #7128 has buyers present, and seems heckin' useful.
- This PR (for #5895) does not (yet); may still be useful tho, so worth not closing?

@rpoyner-tri You agree with this sentiment?
</blockquote></details>

I've assigned myself as feature reviewer, as I'd like to buy this.

---

**Review by avalenzu — COMMENTED** [review]
:lgtm: +@ggould-tri for platform review.

> Reviewed 4 of 4 files at r1.
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/14404#-:-MNa4LoJ1-llaHZH5uFD:b-u268)* status: 1 unresolved discussion, LGTM missing from assignee ggould-tri(platform) (waiting on @EricCousineau-TRI and @ggould-tri)

**Review by sherm1 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/14404)* status: 1 unresolved discussion, LGTM missing from assignee ggould-tri(platform) (waiting on @EricCousineau-TRI and @ggould-tri)

*[a discussion](https://reviewable.io/reviews/robotlocomotion/drake/14404#-MN_IYeCDR0J27_9AM32:-MNa6nIG35TO9GhYudZk:b1c749w) (no related file):*
<details><summary><i>Previously, avalenzu (Andres Valenzuela) wrote…</i></summary><blockquote>

I've assigned myself as feature reviewer, as I'd like to buy this.
</blockquote></details>

Yay, a customer!

---

**Review by EricCousineau-TRI — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/14404)* status: LGTM missing from assignee ggould-tri(platform) (waiting on @ggould-tri)

*[a discussion](https://reviewable.io/reviews/robotlocomotion/drake/14404#-MN_IYeCDR0J27_9AM32:-MNaLVMaBvia0AxtkjjF:brsacyu) (no related file):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

Yay, a customer!
</blockquote></details>

:confetti_ball: Sold!!!

---

**Review by rpoyner-tri — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/14404)* status: LGTM missing from assignee ggould-tri(platform) (waiting on @ggould-tri)

*[a discussion](https://reviewable.io/reviews/robotlocomotion/drake/14404#-MN_IYeCDR0J27_9AM32:-MNcfa3g2HYykT3hwV2N:b-m6u26j) (no related file):*
<details><summary><i>Previously, EricCousineau-TRI (Eric Cousineau) wrote…</i></summary><blockquote>

:confetti_ball: Sold!!!
</blockquote></details>

I guess that answers it. Thanks, gang!

---

**Review by ggould-tri — COMMENTED** [review]
:lgtm:.  Over 80 instances of this pattern in our downstream codebase, so I'm delighted to see this fixed.  Once this is in I'm going to start defecting anonymous systems :-)

> Reviewed 4 of 4 files at r1.
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/14404#-:-MNcgLdn7SJlpTzEypWj:b-dhpyjq)* status: :shipit: complete! all discussions resolved, LGTM from assignees ggould-tri(platform),avalenzu
